Analysis/LeptonID/src/LeptonIDProcessor.cc

```diff
@@ -0,0 +1,374 @@
+    if (_buildTree) {
+        // if I keep this here it magically appears in the aida file :)
+        _tree = new TTree("pfos", "pfos");
+        setupTree();
+    }
+
+    if (_evalMVA) {
+        setupMVAReader();
+    }
+
+}
+
+void LeptonIDProcessor::clear()
+{
+    _truePDG = 0;
+    _parentPDG = 0;
+
+    for (auto v : _vars) {
+        v.second = 0.0f;
+    }
+}
+
+void LeptonIDProcessor::processRunHeader(LCRunHeader *run)
+{
+    (void) run;
+}
+
+void LeptonIDProcessor::modifyEvent(LCEvent *evt)
+{
+    streamlog_out(DEBUG) << "process Event called" << std::endl;
+    LCCollection *PandoraPFOs = nullptr;
+    LCCollection *InputPFOs = nullptr;
+    LCCollection *RecoMCTruthLink = nullptr;
+    LCCollection *MCTruthRecoLink = nullptr;
+    LCCollection *PandoraClusters = nullptr;
+
+    try
+    {
+        PandoraPFOs = evt->getCollection(_PandoraPFOsColName);
+        InputPFOs = evt->getCollection(_InputPFOsColName);
+        PandoraClusters = evt->getCollection(_PandoraClustersColName);
+        if (_buildTree) {
+            RecoMCTruthLink = evt->getCollection(_RecoMCTruthLinkName);
+            MCTruthRecoLink = evt->getCollection(_MCTruthRecoLinkName);
+        }
+    }
+    catch (lcio::DataNotAvailableException const& e)
+    {
+        streamlog_out(WARNING) << "atleast one collection not available" << std::endl;
+        streamlog_out(WARNING) << e.what() << std::endl;
+        return;
+    }
+
+    //LCRelationNavigator RecoMCTruthNavigator(RecoMCTruthLink);
+    //LCRelationNavigator MCTruthRecoNavigator(MCTruthRecoLink);
+    std::unique_ptr<LCRelationNavigator> RecoMCTruthNavigator;
+    std::unique_ptr<LCRelationNavigator> MCTruthRecoNavigator;
+    if (_buildTree) {
+        RecoMCTruthNavigator = std::make_unique<LCRelationNavigator>(RecoMCTruthLink);
+        MCTruthRecoNavigator = std::make_unique<LCRelationNavigator>(MCTruthRecoLink);
+    }
+
+    PIDHandler pidh(PandoraPFOs);
+    int algoID = pidh.addAlgorithm(_pidname, {"mva_e", "mva_mu", "mva_had"});
+    int dEdxAlgId = pidh.getAlgorithmID(_dEdxname);
+    int dEdxDist_e_Id = pidh.getParameterIndex(dEdxAlgId, "electron_dEdxdistance");
+
+    int n = InputPFOs->getNumberOfElements();
+
+    for (int i = 0; i < n; i++)
+    {
+        streamlog_out(DEBUG) << "pfo loop: " << i << std::endl;
+        auto *p = dynamic_cast<ReconstructedParticle *>(InputPFOs->getElementAt(i));
+
+        // We only want to look at charged particles with exactly one track and cluster for now
+        if (abs(p->getCharge()) != 1.0) {
+            continue;
+        }
+
+        auto tracks = p->getTracks();
+        int nTracks = tracks.size();
+        if (nTracks != 1) {
+            continue;
+        }
+
+        auto clusters = p->getClusters();
+        int nClusters = clusters.size();
+        if (nClusters != 1) {
+            continue;
+        }
+
+        if (_buildTree) {
+            // TODO: maybe use PIDUtil getMCParticle instead
```

Comment:
Why not do it?

Reply:
Just need to refactor `getMCParticle` a bit for this to be able to pass the weight cuts, but now is the right moment to do that so I will :)